clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/DefaultIngestionBackend.java

```diff
@@ -50,13 +65,79 @@ public void startConsumption(VeniceStoreVersionConfig storeConfig, int partition
       }
       return storageEngineAtomicReference;
     });
-    LOGGER.info(
-        "Retrieved storage engine for store {} partition {}. Starting consumption in ingestion service",
-        storeVersion,
-        partition);
-    getStoreIngestionService().startConsumption(storeConfig, partition);
-    LOGGER
-        .info("Completed starting consumption in ingestion service for store {} partition {}", storeVersion, partition);
+
+    CompletionStage<Void> bootstrapFuture =
+        bootstrapFromBlobs(storeAndVersion.getFirst(), storeAndVersion.getSecond().getNumber(), partition);
+
+    bootstrapFuture.whenComplete((result, throwable) -> {
+      LOGGER.info(
+          "Retrieved storage engine for store {} partition {}. Starting consumption in ingestion service",
+          storeVersion,
+          partition);
+      getStoreIngestionService().startConsumption(storeConfig, partition);
+      LOGGER.info(
+          "Completed starting consumption in ingestion service for store {} partition {}",
+          storeVersion,
+          partition);
+    });
+  }
+
+  /**
+   * Bootstrap from the blobs from another source (like another peer). If it fails (due to the 30-minute timeout or
+   * any exceptions), it deletes the partially downloaded blobs, and eventually falls back to bootstrapping from Kafka.
+   * Blob transfer should be enabled to boostrap from blobs, and it currently only supports batch-stores.
+   */
+  private CompletionStage<Void> bootstrapFromBlobs(Store store, int versionNumber, int partitionId) {
+    if (!store.isBlobTransferEnabled() || store.isHybrid()) {
+      return CompletableFuture.completedFuture(null);
+    }
+
+    String storeName = store.getName();
+    CompletionStage<InputStream> p2pTransfer = blobTransferManager.get(storeName, versionNumber, partitionId);
+    CompletableFuture<Void> result = new CompletableFuture<>();
+
+    blobTransferScheduler.schedule(() -> {
+      if (!result.isDone()) {
```

Comment:
You are using this `result` incorrectly and we don't need it at all. This `result` has nothing to do with our p2p future. We should chain this with `p2pTransfer`